cpp/inc/bond/protocol/protobuf_binary_writer.h

```diff
@@ -0,0 +1,681 @@
+        struct type : OutputCounter // Must be a new type and not an alias.
+        {};
+    };
+
+
+    template <typename BufferT>
+    class ProtobufBinaryWriter
+    {
+        class Pass1
+        {
+        public:
+            Pass1(ProtobufBinaryWriter* writer)
+                : _writer(writer)
+            {}
+
+            ~Pass1()
+            {
+                _writer->_it = nullptr;
+            }
+
+        private:
+            ProtobufBinaryWriter* _writer;
+        };
+
+        using Counter = ProtobufBinaryCounter::type;
+
+    public:
+        using Buffer = BufferT;
+        using Reader = ProtobufBinaryReader<Buffer>;
+        using Pass0 = ProtobufBinaryWriter<Counter>;
+
+        explicit ProtobufBinaryWriter(Buffer& output)
+            : _output{ output },
+              _it{ nullptr }
+        {}
+
+        template <typename T>
+        ProtobufBinaryWriter(Counter& output, const ProtobufBinaryWriter<T>& /*pass1*/)
+            : _output{ output },
+              _it{ nullptr }
+        {}
+
+        bool NeedPass0() const
+        {
+            return _it == nullptr;
+        }
+
+        Pass1 WithPass0(Pass0& pass0)
+        {
+            _it = pass0._lengths.begin();
+            return this;
+        }
+
+        void WriteVersion()
+        {}
+
+        //
+        // Write methods
+        //
+        void WriteStructBegin(const Metadata& /*metadata*/, bool base)
+        {
+            if (base)
+            {
+                detail::proto::NotSupportedException("Inheritance");
+            }
+
+            if (!_fields.empty())
+            {
+                const auto& info = _fields.top(std::nothrow);
+
+                WriteTag(info.has_element
+                    ? info.element.value.tag
+                    : detail::proto::MakeTag(info.field.id, WireType::LengthDelimited));
+            }
+
+            StructLengthBegin();
+        }
+
+        void WriteStructEnd(bool base = false)
+        {
+            BOOST_VERIFY(!base);
+
+            LengthEnd();
+
+            if (!_fields.empty())
+            {
+                auto& info = _fields.top(std::nothrow);
+
+                if (info.has_element && info.element.map_tag != 0)
+                {
+                    KeyValueEnd(info);
+                }
+            }
+        }
+
+        template <typename T>
+        void WriteField(uint16_t id, const Metadata& metadata, const T& value)
+        {
+            WriteBasic(id, metadata, value);
+        }
+
+        void WriteFieldBegin(BondDataType type, uint16_t id, const Metadata& metadata)
+        {
+            FieldInfo info;
+            info.field.id = id;
+            info.field.metadata = &metadata;
+            info.has_element = false;
+            info.is_list = (type == BT_LIST);
+
+            _fields.push(info);
+        }
+
+        void WriteFieldBegin(BondDataType type, uint16_t id)
+        {
+            WriteFieldBegin(type, id, schema<Unknown>::type::metadata);
```

Comment:
Need to skip the unknown fields during transcoding instead of writing them with default encoding which might be wrong.